File: java/src/main/java/net/snowflake/floe/FloeParameterSpec.java
package net.snowflake.floe;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;

import static net.snowflake.floe.BaseSegmentProcessor.headerTagLength;

public class FloeParameterSpec {
  public static final FloeParameterSpec GCM256_SHA384_4K = new FloeParameterSpec(Aead.AES_GCM_256, Hash.SHA384, 4 * 1024, 32);
  public static final FloeParameterSpec GCM256_SHA384_1M = new FloeParameterSpec(Aead.AES_GCM_256, Hash.SHA384, 1024 * 1024, 32);

  private final Aead aead;
  private final Hash hash;
  private final int encryptedSegmentLength;
  private final int floeIvLength;
  private final Integer keyRotationModuloOverride;
  private final Long maxSegmentNumberOverride;
  private final byte[] encodedParams;

  public FloeParameterSpec(Aead aead, Hash hash, int encryptedSegmentLength, int floeIvLength) {
    this(
        aead,
        hash,
        encryptedSegmentLength,
        floeIvLength,
        null,
        null);
  }

  FloeParameterSpec(
      Aead aead,
      Hash hash,
      int encryptedSegmentLength,
      int floeIvLength,
      Integer keyRotationModuloOverride,
      Long maxSegmentNumberOverride) {
    this.aead = aead;
    this.hash = hash;
    this.encryptedSegmentLength = encryptedSegmentLength;
    this.floeIvLength = floeIvLength;
    this.keyRotationModuloOverride = keyRotationModuloOverride;
    this.maxSegmentNumberOverride = maxSegmentNumberOverride;
    if (encryptedSegmentLength <= 0) {
      throw new IllegalArgumentException("encryptedSegmentLength must be > 0");
    }
    if (floeIvLength != 32) {
      throw new IllegalArgumentException("Currently, floeIvLength must be equal to 32");
    }
    this.encodedParams = paramEncode();
  }

  public static FloeParameterSpec fromHeader(byte[] header) {
    return paramDecode(header);
  }

  private byte[] paramEncode() {
    ByteBuffer result = ByteBuffer.allocate(10).order(ByteOrder.BIG_ENDIAN);
    result.put(aead.getId());
    result.put(hash.getId());
    result.putInt(encryptedSegmentLength);
    result.putInt(floeIvLength);
    return result.array();
  }

  private static FloeParameterSpec paramDecode(byte[] header) {
    ByteBuffer headerBuf = ByteBuffer.wrap(header).order(ByteOrder.BIG_ENDIAN);
    Aead aead = Aead.from(headerBuf.get());
    Hash hash = Hash.from(headerBuf.get());
    int encryptedSegmentLength = headerBuf.getInt();
    int floeIvLength = headerBuf.getInt();
    return new FloeParameterSpec(aead, hash, encryptedSegmentLength, floeIvLength);
  }

  public Aead getAead() {
    return aead;
  }

  public Hash getHash() {
    return hash;
  }

  public int getFloeIvLength() {
    return floeIvLength;
  }

  public int getEncryptedSegmentLength() {
    return encryptedSegmentLength;
  }

  public int getPlainTextSegmentLength() {
    return encryptedSegmentLength - aead.getIvLength() - aead.getAuthTagLength() - Floe.SEGMENT_SIZE_MARKER_LENGTH;
  }

  int getKeyRotationMask() {
    return Optional.ofNullable(keyRotationModuloOverride).orElse(aead.getKeyRotationMask());
  }

  long getMaxSegmentNumber() {
    return Optional.ofNullable(maxSegmentNumberOverride).orElse(aead.getMaxSegmentNumber());
  }

  public byte[] getEncodedParams() {
    return encodedParams.clone();
  }

  public int getHeaderSize() {
    return encodedParams.length + floeIvLength + headerTagLength;
  }

  @Override
  public boolean equals(Object o) {
    if (o == null || getClass() != o.getClass()) return false;
    FloeParameterSpec that = (FloeParameterSpec) o;
    return encryptedSegmentLength == that.encryptedSegmentLength
        && floeIvLength == that.floeIvLength
        && aead == that.aead
        && hash == that.hash
        && Objects.equals(keyRotationModuloOverride, that.keyRotationModuloOverride)
        && Objects.equals(maxSegmentNumberOverride, that.maxSegmentNumberOverride)
        && Objects.deepEquals(encodedParams, that.encodedParams);
  }

  @Override
  public int hashCode() {
    return Objects.hash(aead, hash, encryptedSegmentLength, floeIvLength, keyRotationModuloOverride, maxSegmentNumberOverride, Arrays.hashCode(encodedParams));
  }

  @Override
  public String toString() {
    return String.format("%s{encryptedSegmentLength=%s, floeIvLength=%s, aeadId=%s, hashId=%s}", this.getClass().getCanonicalName(), encryptedSegmentLength, floeIvLength, aead.getId(), hash.getId());
  }
}

File: cpp/src/FloeParameterSpec.cpp
#include <utility>
#include <algorithm>

#include "floe/FloeParameterSpec.hpp"
#include "floe/FloeException.hpp"

namespace floe {

static constexpr int SEGMENT_SIZE_MARKER_LENGTH = 4;
static constexpr int HEADER_TAG_LENGTH = 32;

const FloeParameterSpec FloeParameterSpec::GCM256_SHA384_4K = FloeParameterSpec(
    Aead::fromType(AeadType::AES_GCM_256),
    Hash::fromType(HashType::SHA384),
    4 * 1024,
    32
);

const FloeParameterSpec FloeParameterSpec::GCM256_SHA384_1M = FloeParameterSpec(
    Aead::fromType(AeadType::AES_GCM_256),
    Hash::fromType(HashType::SHA384),
    1024 * 1024,
    32
);

FloeParameterSpec::FloeParameterSpec(
    const Aead& aead,
    const Hash& hash,
    const int encryptedSegmentLength,
    const int floeIvLength)
    : FloeParameterSpec(aead, hash, encryptedSegmentLength, floeIvLength,
                        std::nullopt, std::nullopt) {
}

FloeParameterSpec::FloeParameterSpec(
    Aead aead,
    Hash hash,
    const int encryptedSegmentLength,
    const int floeIvLength,
    const std::optional<int> keyRotationModuloOverride,
    const std::optional<uint64_t> maxSegmentNumberOverride)
    : aead_(std::move(aead)),
      hash_(std::move(hash)),
      encryptedSegmentLength_(encryptedSegmentLength),
      floeIvLength_(floeIvLength),
      keyRotationModuloOverride_(keyRotationModuloOverride),
      maxSegmentNumberOverride_(maxSegmentNumberOverride) {
    
    if (encryptedSegmentLength <= 0) {
        throw FloeException("encryptedSegmentLength must be > 0");
    }
    if (floeIvLength != 32) {
        throw FloeException("Currently, floeIvLength must be equal to 32");
    }
    
    encodedParams_ = paramEncode();
}

std::vector<uint8_t> FloeParameterSpec::paramEncode() const {
    std::vector<uint8_t> result(10);
    
    result[0] = static_cast<uint8_t>(aead_.getType());
    result[1] = hash_.getId();

    const uint32_t encSegLenBE = __builtin_bswap32(static_cast<uint32_t>(encryptedSegmentLength_));
    const auto* encSegBytes = reinterpret_cast<const uint8_t*>(&encSegLenBE);
    std::copy_n(encSegBytes, 4, &result[2]);

    const uint32_t ivLenBE = __builtin_bswap32(static_cast<uint32_t>(floeIvLength_));
    const auto* ivLenBytes = reinterpret_cast<const uint8_t*>(&ivLenBE);
    std::copy_n(ivLenBytes, 4, &result[6]);
    
    return result;
}

int FloeParameterSpec::getPlainTextSegmentLength() const {
    return encryptedSegmentLength_ - aead_.getIvLength() - 
           aead_.getAuthTagLength() - SEGMENT_SIZE_MARKER_LENGTH;
}

int FloeParameterSpec::getKeyRotationMask() const {
    return keyRotationModuloOverride_.value_or(aead_.getKeyRotationMask());
}

uint64_t FloeParameterSpec::getMaxSegmentNumber() const {
    return maxSegmentNumberOverride_.value_or(aead_.getMaxSegmentNumber());
}

std::vector<uint8_t> FloeParameterSpec::getEncodedParams() const {
    return encodedParams_;
}

int FloeParameterSpec::getHeaderSize() const {
    return static_cast<int>(encodedParams_.size()) + floeIvLength_ + HEADER_TAG_LENGTH;
}

}
